components/engine_traits/src/tablet.rs

```diff
@@ -0,0 +1,398 @@
+    // region_id, suffix -> tablet
+    tablets: Mutex<HashMap<u64, CachedTablet<EK>>>,
+    tombstone: Mutex<Vec<(u64, u64)>>,
+    factory: Box<dyn TabletFactory<EK>>,
+    root: PathBuf,
+}
+
+pub struct TabletRegistry<EK> {
+    // One may consider to add cache to speed up access. But it also makes it more
+    // difficult to gc stale cache.
+    tablets: Arc<TabletRegistryInner<EK>>,
+}
+
+impl<EK> Clone for TabletRegistry<EK> {
+    fn clone(&self) -> Self {
+        Self {
+            tablets: self.tablets.clone(),
+        }
+    }
+}
+
+unsafe impl<EK: Send> Send for TabletRegistry<EK> {}
+unsafe impl<EK: Send> Sync for TabletRegistry<EK> {}
+
+impl<EK> TabletRegistry<EK> {
+    pub fn new(factory: Box<dyn TabletFactory<EK>>, path: impl Into<PathBuf>) -> Result<Self> {
+        let root = path.into();
+        std::fs::create_dir_all(&root)?;
+        Ok(TabletRegistry {
+            tablets: Arc::new(TabletRegistryInner {
+                tablets: Mutex::new(HashMap::default()),
+                factory,
+                root,
+                tombstone: Mutex::default(),
+            }),
+        })
+    }
+
+    pub fn tablet_name(&self, prefix: &str, id: u64, suffix: u64) -> String {
+        format!("{}{}_{}", prefix, id, suffix)
+    }
+
+    pub fn tablet_root(&self) -> &Path {
+        &self.tablets.root
+    }
+
+    pub fn tablet_path(&self, id: u64, suffix: u64) -> PathBuf {
+        let name = self.tablet_name("", id, suffix);
+        self.tablets.root.join(name)
+    }
+
+    /// Gets a tablet.
+    pub fn get(&self, id: u64) -> Option<CachedTablet<EK>>
+    where
+        EK: Clone,
+    {
+        let tablets = self.tablets.tablets.lock().unwrap();
+        tablets.get(&id).cloned()
+    }
+
+    /// Gets a tablet, create a default one if it doesn't exist.
+    pub fn get_or_default(&self, id: u64) -> CachedTablet<EK>
+    where
+        EK: Clone,
+    {
+        let mut tablets = self.tablets.tablets.lock().unwrap();
+        tablets
+            .entry(id)
+            .or_insert_with(|| CachedTablet::new(None))
+            .clone()
+    }
+
+    pub fn tablet_factory(&self) -> &dyn TabletFactory<EK> {
+        self.tablets.factory.as_ref()
+    }
+
+    pub fn remove(&self, id: u64) {
+        self.tablets.tablets.lock().unwrap().remove(&id);
+    }
+
+    /// Load the tablet and set it as the latest.
+    ///
+    /// If the tablet doesn't exist, it will create an empty one.
+    pub fn load(&self, id: u64, suffix: u64, create: bool) -> Result<CachedTablet<EK>>
+    where
+        EK: Clone,
+    {
+        let path = self.tablet_path(id, suffix);
+        if !create && !self.tablets.factory.exists(&path) {
+            return Err(Error::Other(box_err!(
+                "tablet ({}, {:?}) doesn't exist",
+                id,
+                suffix
+            )));
+        }
+        let tablet = self.tablets.factory.open_tablet(id, Some(suffix), &path)?;
+        let mut cached = self.get_or_default(id);
```

Comment:
What if `cached` already contains (id, suffix)?

Reply:
Then it will return error at L220 as the tablet is locked.